src/grisp_connect_client.erl

```diff
@@ -60,68 +68,19 @@ terminate(_Reason, _State, _Data) -> ok.
 
 code_change(_Vsn, State, Data, _Extra) -> {ok, State, Data}.
 
-callback_mode() -> [handle_event_function, state_enter].
+callback_mode() -> [state_functions, state_enter].
 
 %%% STATE CALLBACKS ------------------------------------------------------------
 
-% Generic events handling
-handle_event(cast, connect, State, _Data) when State =/= idle ->
-    keep_state_and_data;
-
-handle_event({call, From}, is_connected, connected, _) ->
-    {keep_state_and_data, [{reply, From, true}]};
-handle_event({call, From}, is_connected, _, _) ->
-    {keep_state_and_data, [{reply, From, false}]};
-
-handle_event({call, From}, {request, _, _, _}, State, _Data)
-when State =/= connected ->
-    {keep_state_and_data, [{reply, From, {error, disconnected}}]};
-handle_event({call, From}, {request, Method, Type, Params}, connected,
-            #data{requests = Requests} = Data) ->
-    {ID, Payload} = grisp_connect_api:request(Method, Type, Params),
-    grisp_connect_ws:send(Payload),
-    NewRequests = Requests#{ID => From},
-    {keep_state,
-     Data#data{requests = NewRequests},
-     [{{timeout, ID}, request_timeout(), request}]};
-
-handle_event(cast, {handle_message, Payload}, connected,
-            #data{requests = Requests} = Data) ->
-    Replies = grisp_connect_api:handle_msg(Payload),
-    % A reduce operation is needed to support jsonrpc batch comunications
-    case Replies of
-        [] ->
-            keep_state_and_data;
-        [{request, Response}] -> % Response for a GRiSP.io request
-            grisp_connect_ws:send(Response),
-            keep_state_and_data;
-        [{response, ID, Response}] -> % GRiSP.io response
-            {OtherRequests, Actions} = dispatch_response(ID, Response, Requests),
-            {keep_state, Data#data{requests = OtherRequests}, Actions}
-    end;
-
-handle_event({timeout, ID}, request, connected,
-            #data{requests = Requests} = Data) ->
-    Caller = maps:get(ID, Requests),
-    {keep_state,
-        Data#data{requests = maps:remove(ID, Requests)},
-        [{reply, Caller, {error, timeout}}]};
-
-handle_event({call, From}, _, State, Data) ->
-    {keep_state, Data, [{reply, From, {bad_client_state, State}}]};
-
-% STATE MACHINE Transitions
-
-% IDLE
-handle_event(enter, _OldState, idle, _Data) ->
+idle(enter, _OldState, _Data) ->
     keep_state_and_data;
-handle_event(cast, connect, idle, Data) ->
+idle(cast, connect, Data) ->
     {next_state, waiting_ip, Data};
+?HANDLE_COMMON.
 
-% WAITING_IP
-handle_event(enter, _OldState, waiting_ip, Data) ->
+waiting_ip(enter, _OldState, Data) ->
     {next_state, waiting_ip, Data, [{state_timeout, 0, retry}]};
```

Comment:
Well this line wasn't changed, but I noticed: You can use `{keep_state_and_data, {state_timeout, 0 , retry}}` here.